src/can.rs

```diff
@@ -0,0 +1,111 @@
+//! Controller Area Network
+
+use core::convert::TryInto;
+use core::convert::TryFrom;
+
+use nb;
+
+/// A type that can either be `BaseId` or `ExtendedId`
+#[cfg(feature = "unproven")]
+pub trait Id: Sized {
+    /// The BaseId variant
+    type BaseId: BaseId + Into<Self> + TryFrom<Self>;
+
+    /// The ExtendedId variant
+    type ExtendedId: ExtendedId + Into<Self> + TryFrom<Self>;
+}
+
+/// A Can (11-bit) ID
+#[cfg(feature = "unproven")]
+pub trait BaseId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Extended (28-bit) ID
+#[cfg(feature = "unproven")]
+pub trait ExtendedId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Frame
+#[cfg(feature = "unproven")]
+pub trait Frame {
+    /// The Id type of this Frame
+    type Id: Id;
+
+    /// Returns the Can-ID
+    fn id(&self) -> Self::Id;
+
+    /// Returns `Some(Data)` if data frame.
+    /// Returns `None` if remote frame.
```

Comment:
This seems like a non-obvious way to encode remote frames.  If I asked a random programmer "What would you guess is the difference between a CAN frame with None for the data, and a CAN frame with Some(empty_slice) as the data?"  I'm skeptical they'd guess "Whether or not its a remote frame!"

Or to voice my concern from a different angle, Option means "may not exist", where as RTR means "I would like you to do a thing", and those aren't well connected.

Reply:
I agree. How can we change this to the better?

Returning a len 0 slice for a remote frame is worse i believe. At the same time we will need to support frame encodings where it's not statically known if they are remote or data frames (from the discussion in the last PR)

Would it be sufficient to introduce `is_remote_frame(&self) -> bool` and `is_data_frame(&self) -> bool` as well?

Reply:
Why is a len 0 slice worse?  On the wire the length field is indeed zero and zero bytes come before the checksum.  It seems like a great modeling of what is actually happening on the bus.

I'm not sure what you mean wrt not knowing if a frame is remote or data, it's a required bit, it will always be present and always be either 0 or 1...?  What's a situation where you might not know?

But yes, `is_remote_frame(&self) -> bool` seems like a fine solution.

Reply:
I think you'll find it challenging to come up with an API that is natural for both static and dynamic typing.  Static typing would have it that the id types are different in some way, and dynamic would have it that everything from either id type is on both.

But I hear what you're saying about the conceptual difference.  And you made a good point in the other PR, namely, that the compiler should enforce that RTR frames have no data.  I would say, though, that this conceptual difference is a bit idealistic.  Obviously it's best if invalid configurations can't be represented, but what if invalid configurations are what other devices on the bus are sending?  Should the HAL prevent communication with such devices (that are, generally, not within the user's control, eg the ECUs in a car)?

Reply:
Due to the nature of this crate is essential that it works for all reasonable representations of the can frame. 

>  I would say, though, that this conceptual difference is a bit idealistic.

Sure, but I would also argue that it's an idiomatic use of Rusts typesystem.

>  Obviously it's best if invalid configurations can't be represented, but what if invalid configurations are what other devices on the bus are sending? Should the HAL prevent communication with such devices (that are, generally, not within the user's control, eg the ECUs in a car)?

The DLC field in a remote frame doesn't mean "how long is the data field for this message" it means "How much data do i want to receive". Therefore the CAN controller would never be able to assemble such frame and this would never be an issue. But even if it were, I would almost always go for the fail fast and loudly approach and not accept a malformed Frame.

Reply:
```rust
enum ((I don't know what this should be named)) {
    Remote,
    Data(&[u8])
}
```
This would be an idiomatic solution that clearly represents what this frame is.